components/object_store/src/multipart.rs

```diff
@@ -0,0 +1,258 @@
+    tasks: FuturesUnordered<BoxedTryFuture<(usize, UploadPart)>>,
+    /// Maximum number of upload tasks to run concurrently
+    max_concurrency: usize,
+    /// Buffer that will be sent in next upload.
+    current_buffer: Vec<u8>,
+    /// Size of a part in bytes, size of last part may less than `part_size`.
+    part_size: usize,
+    /// Index of current part
+    current_part_idx: usize,
+    /// The completion task
+    completion_task: Option<BoxedTryFuture<()>>,
+}
+
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl,
+{
+    pub fn new(inner: T, max_concurrency: usize, part_size: usize) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            completed_parts: Vec::new(),
+            tasks: FuturesUnordered::new(),
+            max_concurrency,
+            current_buffer: Vec::new(),
+            part_size,
+            current_part_idx: 0,
+            completion_task: None,
+        }
+    }
+
+    pub fn poll_tasks(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Result<(), io::Error> {
+        if self.tasks.is_empty() {
+            return Ok(());
+        }
+        while let Poll::Ready(Some(res)) = self.tasks.poll_next_unpin(cx) {
+            let (part_idx, part) = res?;
+            let total_parts = self.completed_parts.len();
+            self.completed_parts
+                .resize(std::cmp::max(part_idx + 1, total_parts), None);
+            self.completed_parts[part_idx] = Some(part);
+        }
+        Ok(())
+    }
+}
+
+/// [ Note: Methods in this impl are added by ceresdb, not included in the
+/// `object_store` crate ].
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    /// Send all buffer data to object store in final stage.
+    fn final_flush_buffer(mut self: Pin<&mut Self>) {
+        let part_size = self.part_size;
+        while !self.current_buffer.is_empty() {
+            let size = part_size.min(self.current_buffer.len());
+            let out_buffer = self.current_buffer.drain(0..size).collect::<Vec<_>>();
+
+            let inner = Arc::clone(&self.inner);
+            let part_idx = self.current_part_idx;
+            self.tasks.push(Box::pin(async move {
+                let upload_part = inner.put_multipart_part(out_buffer, part_idx).await?;
+                Ok((part_idx, upload_part))
+            }));
+            self.current_part_idx += 1;
+        }
+    }
+
+    /// Send buffer data to object store in write stage.
+    fn flush_buffer(mut self: Pin<&mut Self>) {
+        let part_size = self.part_size;
+
+        // We will continuously submit tasks until size of the buffer is smaller than
+        // `part_size`.
+        while self.current_buffer.len() >= part_size {
+            let out_buffer = self.current_buffer.drain(0..part_size).collect::<Vec<_>>();
+            let inner = Arc::clone(&self.inner);
+            let part_idx = self.current_part_idx;
```

Comment:
Add a method called `fetch_add_part_idx` which will do: 
```rust
let part_idx = self.current_part_idx;
self.current_part_idx += 1;
part_idx
```
After that, use it here and in `final_flush_buffer`.